### PR TITLE
JP-3510: Use all groups to flag bad IRS2 reference pixels

### LINKED ISSUE #8195: Use all groups to flag bad IRS2 reference pixels
_Issue [JP-3510](https://jira.stsci.edu/browse/JP-3510) was created on JIRA by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke):_

In JP-3437, it was pointed out that the current bad reference pixel flagging algorithm frequently misses intermittent bad pixels because it only checks the last group of the last integration.  If a reference pixel happens to be okay in that group but bad in another, it will not be caught, but it will still impact the outcome.

In addition, bad pixels marked in the DQ table in the IRS2 reference file are replaced with 0.0, to be interpolated over later.  Bad pixels marked by the intermittent flagging algorithm are replaced with their nearest good value, no matter how far away the replacement is.  These replacement options should be tested to see which is better – the better one should be used for both kinds of bad reference pixels.

### PR DESCRIPTION
**Pull request by melanieclarke**
Resolves [JP-3510](https://jira.stsci.edu/browse/JP-3510)

Closes #8195 

This PR improves on-the-fly bad reference pixel flagging by using all groups from all integrations to compile mean and standard deviation statistics.  This should catch reference pixels that are intermittently bad and happened to be okay in the last group.

Additionally, this PR moves the flagging outside of the IRS2 correction method, so that it can be applied before any other reference pixel handling is done, and updates the pixel replacement method for both newly flagged bad pixels and known bad pixels.  Replacement values come from the nearest reference groups, respecting parity if possible; 0.0 is used as a replacement value only as a last resort.

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -28,6 +28,13 @@ extract_1d
 - Fixed a bug in the calling of optional MIRI MRS 1d residual fringe
   correction that could cause defringing to fail in some cases. [#8180]
 
+refpix
+------
+
+- Modify IRS2 bad reference pixel flagging to consider values from all
```
Comment (hbushouse) [inline CHANGES.rst:34]:
```suggestion
- Modify NIRSpec IRS2 bad reference pixel flagging to consider values from all
```

File: jwst/refpix/irs2_subtract_reference.py [hunk 2]
```diff
@@ -74,6 +72,7 @@ def correct_model(input_model, irs2_model,
     # Only copy in SCI data array for now; that's all we need. The rest
```
Comment (hbushouse) [inline jwst/refpix/irs2_subtract_reference.py:72]: Might want to update this comment, now that both SCI and PIXELDQ are being copied.

**Review by hbushouse — COMMENTED** [review]
Code updates look fine. Just a couple minor comments.

**Comment by hbushouse** [comment]
Have you done a regression test run or do you not have permission to set those up on our Jenkins server?

File: CHANGES.rst (hunk 1 above)
Comment (melanieclarke) [reply CHANGES.rst]: Updated

File: jwst/refpix/irs2_subtract_reference.py (hunk 2 above)
Comment (melanieclarke) [reply jwst/refpix/irs2_subtract_reference.py]: Done, thanks for pointing that out.

**Comment by melanieclarke** [comment]
@hbushouse - I have permission to see regression tests but not launch them.

**Comment by hbushouse** [comment]
Regtest run started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1164

**Comment by hbushouse** [comment]
Regression test results look good. Several tests involving NIRSpec IRS2 data that show expected differences starting at the refpix step, as well as several unrelated differences due to other recent mergers into the master branch.

**Pull request merged**